### ISSUE TITLE
Remove duplicated dependencies

### ISSUE THREAD

**Issue by YuhanLiin** [rust-ml/linfa#138]
Calling `cargo tree -d --workspace` at the root of the repo reveals all of Linfa's duplicated dependencies. These are dependencies that have the same name but different versions, leading to bloat. We want to eliminate as many duplicated dependencies as possible by ensuring that multiple instances of the same crate in the dependency tree have the same version. We should also prioritize the non-dev dependencies since they affect the end users.

**Comment by bytesnake** [comment]
I investigated a bit and came up with these different dependency mismatches:
 * `cfg-if` caused by `reqwest`, opened an issue here https://github.com/davidMcneil/mnist/issues/7
 * `generic-array` caused by `ndarray-npy` which we can bump once we moved to `ndarray = 0.15`
 * `getrandom`, `rand`, `rand-chacha`, `rand_core` caused by using `rand = 0.8` in `linfa`, we could downgrade to `0.7` here or wait for the ndarray bump
 * `hashbrown` caused by an old version of `hnsw`. This also requires a bump in `ndarray-stats/indexmap/hashbrown` dependency to be on the latest version, opened a PR https://github.com/bluss/indexmap/pull/193
 * `itertools` caused by inconsistency in `criterion`'s dependency graph and old version of `ndarray-stats` (can be updated once we have `ndarray = 0.15`)
 * `num-bigint`, `num-complex` caused by `ndarray-npy` which can only be updated once we bump ndarray itself
 * `pin-project-lite` caused by inconcistency in `reqwest` dependency graph, also depends on https://github.com/davidMcneil/mnist/issues/7
 * `socket2` caused by using `reqwest` and `curl` 
 * `time` caused by `reqwest` as well, which uses an older version compared to `argmin`

the problems are therefore originating from three sources:
 * we need the `ndarray` bump to be consistent to the ecosystem
 * `reqwest` 
 * `indexmap` uses an older version of `hashbrown`

**Comment by YuhanLiin** [comment]
`reqwest` is not a big deal because it only affects dev-dependencies. Does downgrading `rand` in `linfa` break anything?

**Comment by bytesnake** [comment]
> `reqwest` is not a big deal because it only affects dev-dependencies. Does downgrading `rand` in `linfa` break anything?

yes the ndarray bump is the main issue here. Downgrading `rand` requires reverting some function signatures. The dependency tree of `rand` is not really worth the work

**Comment by YuhanLiin** [comment]
After bumping `ndarray` to v0.15, `approx` is still duplicated since `argmin` v0.4.6 uses `approx` v0.5 but `ndarray` uses `approx` v0.4.

**Comment by oojo12** [comment]
Otherwise, the approx bump is coming in v0.16.0 of ndarray

**Comment by YuhanLiin** [comment]
The `approx-0_5` feature should solve the issue. Double-check to make sure.
